MAINT: fix mypy type output

pubpub-zz opened the pull request:
closes #2798

stefan6419846 commented:
I am not completely sure about this - using private members in the public API does not feel right.

pubpub-zz commented:
> I am not completely sure about this - using private members in the public API does not feel right.

Can you please clarify: keys and items are public for _VirtualListImages.
if the issue is _VirtualListImages, should we rename it into VirtualListImages. this may have more impact and will make _VirtualListImages too visible without any advantage:  I would not like to create a new subject of discussion such as the one we had on filters/LZWDecode

However, having a library fully compatible with mypy worth the commit

stefan6419846 commented:
In my opinion, it does not make sense to have a type hint referencing `_VirtualListImages` while this class is both not exposed publicly and even has a leading underscore to indicate it is private - such cases should only occur for the internal API. (`ImageFile` never being available as a public class is a similar issue in my opinion, but at least not prefixed by an underscore - the general issue is the same.)

pubpub-zz commented:
Based on your comments: 
* I renamed `_VirtualListImages` to  `VirtualListImages` 
* I moved `ImageFile` to make it public. 
* I broke the inheritance between `ImageFile` with `File` and made it obsolete(never used out of ImageFile) in order to have a standalone module
* I also moved PIL import in _page to global in order to use `(PIL.Image.)Image` type

stefan6419846 commented:
> I also moved PIL import in _page to global in order to use `(PIL.Image.)Image` type

Isn't this a breaking change for users which do not have *Pillow* installed?

pubpub-zz commented:
> > I also moved PIL import in _page to global in order to use `(PIL.Image.)Image` type
> 
> Isn't this a breaking change for users which do not have _Pillow_ installed?

No: the error is reported only when calling replace (as it used to be) There is no change in behavior

stefan6419846 commented on pypdf/_page.py:
```diff
@@ -301,6 +310,160 @@ def apply_on(
+
+    name: str = ""
+    """
+    FileName as identified within the PDF file
```
```suggestion
    Filename as identified within the PDF file.
```

stefan6419846 commented on pypdf/_page.py:
```diff
@@ -301,6 +310,160 @@ def apply_on(
+
+    data: bytes = b""
+    """
+    data as bytes
```
```suggestion
    Data as bytes.
```

stefan6419846 commented on pypdf/_page.py:
```diff
@@ -301,6 +310,160 @@ def apply_on(
+
+    image: Optional[Image] = None
+    """
+    data as PIL image;
```
```suggestion
    Data as PIL image.
```

stefan6419846 commented on pypdf/_page.py:
```diff
@@ -301,6 +310,160 @@ def apply_on(
+
+    indirect_reference: Optional[IndirectObject] = None
+    """
+    Reference to the Object storing the stream
```
```suggestion
    Reference to the object storing the stream.
```

stefan6419846 commented on pypdf/_page.py:
```diff
@@ -301,6 +310,160 @@ def apply_on(
+
+    def replace(self, new_image: Image, **kwargs: Any) -> None:
+        """
+        Replace the Image with a new PIL image.
```
```suggestion
        Replace the image with a new PIL image.
```

stefan6419846 commented on pypdf/_page.py:
```diff
@@ -301,6 +310,160 @@ def apply_on(
+        from .generic import DictionaryObject, PdfObject
+
+        if self.indirect_reference is None:
+            raise TypeError("Can not update an inline image")
```
```suggestion
            raise TypeError("Cannot update an inline image.")
```

stefan6419846 commented on pypdf/_page.py:
```diff
@@ -301,6 +310,160 @@ def apply_on(
+        if self.indirect_reference is None:
+            raise TypeError("Can not update an inline image")
+        if not hasattr(self.indirect_reference.pdf, "_id_translated"):
+            raise TypeError("Can not update an image not belonging to a PdfWriter")
```
```suggestion
            raise TypeError("Cannot update an image not belonging to a PdfWriter.")
```

stefan6419846 commented on pypdf/_page.py:
```diff
@@ -301,6 +310,160 @@ def apply_on(
+class VirtualListImages(Sequence[ImageFile]):
+    """
+    Provides access to images referenced within a page.
+    One copy only will be returned if used many times in the same page
```
```suggestion
    Only one copy will be returned if the usage is used on the same page multiple times.
```

stefan6419846 commented on pypdf/_page.py:
```diff
@@ -301,6 +310,160 @@ def apply_on(
+    """
+    Provides access to images referenced within a page.
+    One copy only will be returned if used many times in the same page
+    see :func:`PageObject.images` for more details
```
```suggestion
    See :func:`PageObject.images` for more details.
```

stefan6419846 commented on pypdf/_utils.py:
```diff
@@ -525,10 +524,18 @@ def getter(self, method):  # type: ignore  # noqa: ANN001, ANN202
-    indirect_reference: Optional[IndirectObject] = None  # optional ; link to PdfObject
+    name: str = ""
+    """
+    FileName as identified within the PDF file
```
```suggestion
    Filename as identified within the PDF file.
```

stefan6419846 commented on pypdf/_utils.py:
```diff
@@ -525,10 +524,18 @@ def getter(self, method):  # type: ignore  # noqa: ANN001, ANN202
+    """
+    data: bytes = b""
+    """
+    data as bytes
```
```suggestion
    Data as bytes.
```

stefan6419846 commented on pypdf/_utils.py:
```diff
@@ -525,10 +524,18 @@ def getter(self, method):  # type: ignore  # noqa: ANN001, ANN202
+    """
+    indirect_reference: Optional[IndirectObject] = None
+    """
+    Reference to the Object storing the stream
```
```suggestion
    Reference to the object storing the stream.
```

stefan6419846 commented on tests/test_workflows.py:
```diff
@@ -990,6 +990,13 @@ def test_replace_image(tmp_path):
 
+    import pypdf
+
+    pypdf._page.pil_not_imported = True
```
This should use a construct with `finally` to ensure that we always reset the value correctly.

stefan6419846 reviewed (commented):
I have reworded some of the docs.

Additionally, the `ImageFile` and `VirtualListImages` classes are still part of a private module only without being exposed in the `__init__.py` file. We should expose them there to allow users to actually refer to them in their own type hints if required.

pubpub-zz commented:
> Additionally, the `ImageFile` and `VirtualListImages` classes are still part of a private module only without being exposed in the `__init__.py` file. We should expose them there to allow users to actually refer to them in their own type hints if required.

ImageFile and VirtualListImages are not classes that a user should create instances from : as such I see no reason for them to be part of the `__init__.py`

The objective is just to fix mypy and the test report in the issue was positive

stefan6419846 commented:
Okay, let's keep this for now - except for the other remark regarding `finally` which still should be fixed.
